File: BFS/serialize-deserialize.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * This method will be invoked first, you should design your own algorithm 
     * to serialize a binary tree which denote by a root node to a string which
     * can be easily deserialized by your own "deserialize" method later.
     */
    public String serialize(TreeNode root) {
        if (root == null) {
            return "{}";
        }

        // bfs 结果存储数组
        List<TreeNode> nodes = new ArrayList<>();

        // bfs 过度工具
        Queue<TreeNode> queue = new LinkedList<>();
        
        // 宽度优先搜索获得BFS序
        queue.add(root);
        while (! queue.isEmpty()) {
            TreeNode node = queue.poll();

            // node could be null
            nodes.add(node);
            
            if (node == null) {
                // 不需要继续下去了, stop
                continue;
            }
            queue.add(node.left);
            queue.add(node.right);
        }

        // 清除最后的null
        // [1, 2, nul, null, 3, null, null], 清除最后的null, 没有意义
        while (nodes.get(nodes.size() - 1) == null) {
            nodes.remove(nodes.size() - 1);
        }

        // 将BFS序序列化数组转成字符串
        StringBuffer result = new StringBuffer();
        result.append("{");
        for (int i = 0; i < nodes.size(); i++) {
            if (i > 0) {
                result.append(",");
            }
            if (nodes.get(i) == null) {
                result.append("#");
            }
            else {
                result.append(Integer.toString(nodes.get(i).val));
            }
        }
        result.append("}");
        return result.toString();
    }

File: BFS/serialize-deserialize.js
// BFS
const serialize = (root) => {
  const queue = [root];
  let res = [];

  // 不带level的模板
  while (queue.length) {        // queue not empty

    const node = queue.shift(); // 考察出列的节点
    
    if (node) {                 // 是真实节点，带出子节点入列
      res.push(node.val);       // 节点值推入res
      
      queue.push(node.left);    // 子节点入列，不管是不是null节点都入列
      queue.push(node.right);    
    } else {                    // 是null节点，没有子节点入列
      res.push('X');            // X 推入res
    }
  }
  return res.join(',');         // 数组转成字符串
}

// 作者：xiao_ben_zhu
// 链接：https://leetcode-cn.com/problems/serialize-and-deserialize-binary-tree/solution/shou-hui-tu-jie-gei-chu-dfshe-bfsliang-chong-jie-f/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

const deserialize = (data) => {
  if (data == 'X') return null;

  const list = data.split(',');  // 序列化字符串split成数组

  const root = new TreeNode(list[0]); // 获取首项，构建根节点

  // queue里面存放的全部是父亲节点
  const queue = [root];          // 根节点推入队列

  let cursor = 1;                // 初始指向list第二项

  while (cursor < list.length) { // 指针越界，即扫完了序列化字符串
    
    // node 现在暂时是父亲节点
    const node = queue.shift();  // 考察出列的节点

    // 父亲的两个儿子的值取出来
    const leftVal = list[cursor];      // 它的左儿子的值
    const rightVal = list[cursor + 1]; // 它的右儿子的值

    if (leftVal != 'X') {              // 是真实节点
      const leftNode = new TreeNode(leftVal); // 创建左儿子节点
      node.left = leftNode;                   // 左儿子认父亲
      queue.push(leftNode);             // 左儿子角色互换, 也是父亲，入列queue
    }
    if (rightVal != 'X') {
      const rightNode = new TreeNode(rightVal);
      node.right = rightNode;
      queue.push(rightNode);
    }
    
    cursor += 2; // 一次考察一对儿子，指针加2
  }
  return root;  // BFS结束，构建结束，返回根节点
};

// 作者：xiao_ben_zhu
// 链接：https://leetcode-cn.com/problems/serialize-and-deserialize-binary-tree/solution/shou-hui-tu-jie-gei-chu-dfshe-bfsliang-chong-jie-f/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。